Source/Core/Core/Debugger/BranchWatch.cpp

```diff
@@ -0,0 +1,302 @@
+// Copyright 2023 Dolphin Emulator Project
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#include "Core/Debugger/BranchWatch.h"
+
+#include <algorithm>
+#include <cstdio>
+
+#include <fmt/format.h>
+
+#include "Common/Assert.h"
+#include "Common/CommonTypes.h"
+#include "Core/Core.h"
+#include "Core/PowerPC/Gekko.h"
+#include "Core/PowerPC/MMU.h"
+
+namespace Core
+{
+bool BranchWatch::Start()
+{
+  if (m_recording_phase != Phase::Disabled)
+    return false;
+
+  m_collection_v.clear();
+  m_collection_p.clear();
+  m_recording_phase = Phase::Blacklist;
+  m_blacklist_size = 0;
+  m_recording_active = true;
+  return true;
+}
+
+void BranchWatch::Stop()
+{
+  m_recording_active = false;
+  m_recording_phase = Phase::Disabled;
+  m_selection.clear();
+}
+
+bool BranchWatch::Pause()
+{
+  if (m_recording_phase == Phase::Disabled)
+    return false;
+
+  m_recording_active = false;
+  return true;
+}
+
+bool BranchWatch::Resume()
+{
+  if (m_recording_phase == Phase::Disabled)
+    return false;
+
+  m_recording_active = true;
+  return true;
+}
+
+void BranchWatch::Save(const Core::CPUThreadGuard&, std::FILE* file) const
+{
+  if (!CanSave())
+  {
+    ASSERT_MSG(CORE, false, "BranchWatch can not be saved.");
```

Comment:
Looking over this again, Is invoking the assertion handler (which can kill dolphin if the panic handler is disabled) necessary over either returning an error code and letting the frontend handle it (usually with a localized and readable message), or printing an error log message.

Reply:
The BranchWatchDialog already checks `bool Core::BranchWatch::CanSave()` before attempting to save, this was just an extra safety check to future-proof the component against misuse.  Do you think an ASSERT is overkill for this?  I can switch to a log message and return false.

Reply:
In that case (I missed the check in the dialog implementation), I think this is perfectly fine